scipy/io/arff/arffread.py

```diff
@@ -58,168 +64,375 @@ class ArffError(IOError):
+            return cls(name, date_format, datetime_unit)
+        else:
+            return None
+
+    def parse_data(self, data_str):
+        """
+        Parse a value of this type.
+        """
+        date_str = data_str.strip().strip("'").strip('"')
+        if date_str == '?':
+            return np.datetime64('NaT', self.datetime_unit)
+        else:
+            dt = datetime.datetime.strptime(date_str, self.date_format)
+            return np.datetime64(dt).astype(
+                "datetime64[%s]" % self.datetime_unit)
+
+    def __str__(self):
+        return super(DateAttribute, self).__str__() + ',' + self.date_format
+
+
+class RelationalAttribute(Attribute):
+
+    def __init__(self, name):
+        super().__init__(name)
+        self.type_name = 'relational'
+        self.dtype = np.object_
+        self.attributes = []
+        self.dialect = None
+
+    @classmethod
+    def parse_attribute(cls, name, attr_string):
+        """
+        Parse the attribute line if it knows how. Returns the parsed
+        attribute, or None.
+
+        For date attributes, the attribute string would be like
+        'date <format>'.
+        """
+
+        attr_string_lower = attr_string.lower().strip()
+
+        if attr_string_lower[:len('relational')] == 'relational':
+            return cls(name)
+        else:
+            return None
+
+    def parse_data(self, data_str):
+        # Copy-pasted
+        elems = list(range(len(self.attributes)))
+
+        escaped_string = data_str.encode().decode("unicode-escape")
+
+        row_tuples = []
+
+        for raw in escaped_string.split("\n"):
+            row, self.dialect = split_data_line(raw, self.dialect)
 
-def go_data(ofile):
-    """Skip header.
+            row_tuples.append(tuple(
+                [self.attributes[i].parse_data(row[i]) for i in elems]))
 
-    the first next() call of the returned iterator will be the @data line"""
-    return itertools.dropwhile(lambda x: not r_datameta.match(x), ofile)
+        return np.array(row_tuples,
+                        [(a.name, a.dtype) for a in self.attributes])
 
+    def __str__(self):
+        return (super(RelationalAttribute, self).__str__() + '\n\t' +
+                '\n\t'.join(str(a) for a in self.attributes))
 
-#----------------
+
+# -----------------
+# Various utilities
+# -----------------
+def to_attribute(name, attr_string):
+    attr_classes = (NominalAttribute, NumericAttribute, DateAttribute,
+                    StringAttribute, RelationalAttribute)
+
+    for cls in attr_classes:
+        attr = cls.parse_attribute(name, attr_string)
+        if attr is not None:
+            return attr
+
+    raise ParseArffError("unknown attribute %s" % attr_string)
+
+
+def split_data_line(line, dialect=None):
+    delimiters = ",\t"
+
+    # This can not be done in a per reader basis, and relational fields
+    # can be HUGE
+    csv.field_size_limit(int(ctypes.c_ulong(-1).value // 2))
```

Comment:
This looks inefficient and changes global state. If you need to play with `field_size_limit`, then it looks to me like you should set it outside the for-loops that call `split_data_line`, and then reset it to the old value when you're done.

Reply:
I have though about that, and the problem would be that if two threads do that, then the limit would be reset to the old value before one of them finishes, which is undesirable. As this variable is typically only modified with the objective of increasing it, I think that leaving it set to the maximum value makes sense. Of course, the preferred solution would be a per-reader or per-thread way of doing the same thing, but no one is available in the `csv` module.

Reply:
Okay fair enough. not ideal, but I think I'm fine with changing global state in this case.